### DISCUSSION TITLE
[Migration] MongoDB Atlas Sync depracated.

### DISCUSSION THREAD

**Discussion by GeoffreyChen777** [Future-Scholars/paperlib#609] [Development]
Hi folks, @igoogolx @charlieJ107 

Thanks you both again for your consistent contribution to Paperlib.

We have a really big trouble. #607 

MongoDB Atlas announced that they are deprecating the device sync and device sdk. Paperlib highly relies on these two functions.

- **device sdk**: the interface of our database.
- **device sync** cross device data sync.

We need to migrate these two functions before Sep 2025.

As it's a really big change, I wish you could help me with it.

The first step is finding a alternative solution.

Really appreciate.

(We can discuss in either English or Chinese, if you prefer. I believe we both speak Chinese)

**Comment by charlieJ107** [comment]
Shall we keep the current repositories interface and implement them with a RESTful API, so we can keep the sync features and allow the further extension

**Comment by GeoffreyChen777** [comment]
@charlieJ107 I think we can propose our own ideas, and we discuss them, finally we choose the best solution. I remember you have some ideas about the database interface before.

Here I provide some key features in the current Paperlib database that we should consider for the alternative solution:

- **Document-based data schema.** Currently, all data structure in Paperlib like a JSON object. Thus it would be better to still use document-based database, such as mongodb etc.
- **Reactive.** The current Paperlib UI is basically a database dashboard. We show the database with different query conditions in the UI.
  - When we change the database, the UI is refreshed automatically. For example, we change the title of a paper, we just need to update the title in the database. The UI will then show the update without manually refreshing.
  - When we change the query condition, the UI is also refreshed automatically.
  - When users use sync database, they update the database on the device A, the updates should be synced to the database on the device B, and refresh the UI of device B.
- **Easy query.**
  - We have a feature called smart-filter. It basically use the realm query language to filter the database.
  - The advanced searching also use the realm query language.
  - The tags/folders in the sidebar are also based on the realm query. There is a query sentence behind each tag/folder. When a user clicks a tag/folder in the sidebar, we send the corresponding query sentence to the database and get the required data.
- **Sync.** This is a really important feature.
  - Database should be automatically synced between different devices.
  - Local first. Users can still use a local copy of the online database if there is no internet connection. After the user connects the internet, the local copy should be merged with the online database.
  - Automatic conflict handeling.

**Comment by charlieJ107** [comment]
I believe the first issue we need to resolve is determining how much we want to change the current codebase and how many breaking changes we are willing to tolerate. Based on this, I propose two potential solutions with their pros and cons.

#### Option 1: Maintain the existing codebase, build a compatibility layer
The core of this solution is to remove the deprecated Realm features and build a compatibility layer to integrate new data storage and access methods into the current codebase.

**Pros:**
1. **Minimized workload**: Most of the work is focused on the data layer, while other parts of the system remain unchanged.
2. **Existing architecture remains intact**: The plugin system and software architecture can largely remain unchanged, reducing risk.
3. **Lower learning curve**: The development team won’t need to adapt to a new architecture, allowing for faster implementation.

**Cons:**
1. **Technical debt**: The compatibility layer might increase future maintenance complexity and add to the technical debt.
2. **Limited flexibility**: The current architecture’s limitations might hinder feature expansion in the future.
3. **Sync and query challenges**: We will need to implement sync and query language compatibility, which could become a bottleneck.

#### Option 2: Full project refactor, front-end/back-end separation
This solution involves a complete project refactor with a front-end/back-end separation architecture, decoupling data access from other services, and redesigning the system to be more flexible and modern.

**Pros:**
1. **Decoupling architecture**: It provides greater flexibility for future feature expansions and reduces dependencies.
2. **Solves foundational issues**: It addresses deep-rooted system problems, such as self-hosted and team-sharing functionalities.
3. **Long-term growth**: The new architecture lays a foundation for future scalability and performance improvements.
4. **Sync and reactive needs are met**: A back-end service can resolve sync and UI auto-refresh issues effectively.

**Cons:**
1. **Significant workload**: It requires a large-scale refactor, meaning a longer development time and more work.
2. **Breaking changes**: The new architecture may cause incompatibility with the existing system, leading to major changes.
3. **High development cost**: The team will need to learn and adapt to the new architecture, increasing the development difficulty.

#### My Recommendation
I think we need to reach a consensus on our short-term and long-term goals. If the goal is to quickly address the issue, Option 1 can serve as a temporary solution. However, in the long run, I favor Option 2 as it offers more flexibility and future scalability. We should carefully balance the costs and benefits of a full refactor and have thorough discussions on how to decouple existing components and adopt best practices to ensure the robustness of the system.

**Comment by GeoffreyChen777** [comment]
@charlieJ107 

Thanks for your ideas.

As we have one year for this, it would be OK to choose option 2.

And, I have already finished the front-end/back-end separation work in the branch `upgrade-and-change-build-system`. Currently, the electron renderer process is just a pure web page, without any node integration. All services are running in two utility processes, one is for the extension-related services, and the other one is for other services including the database.

Thus, if we choose option 2, we will still focus mainly on implementing the database.

if we decide to implement the database part by ourselves, I think maybe we can mix option 1 and option 2:
- based on the current database service interface, we implement a compatibility layer to connect to some databases. I think we don't need to implement a database from scratch. It's too complicated.
- we consider some future features, such as team-sharing, if the current interface is not the optimal design, we change it. Otherwise, we keep it.

**Comment by GeoffreyChen777** [comment]
I'm investigating this database: https://rxdb.info/

It's an open-source database with a sync function.

**Comment by charlieJ107** [comment]
Thank you for your great work on refactoring and decoupling the front end and back end in the new branch. I think we can now focus on designing a scalable and flexible data access architecture. My suggestion is to refer to the common practice of Apps on mobile devices and adopt a two-layer data architecture with multiple parallel Data Sources and a unified Data Repository. Taking the Paper data as an example, there is usually a unified `PaperRepository` to provide data to other components in the application. In `PaperRepository`, a `LocalDataSource `and a `RemoteDataSource` are maintained. For `LocalDataSource`, the common implementation on Android Apps is to use the local Sqlite database, while `RemoteDatasource` uses the HTTP or other protocols to access the API on the cloud to implement data interaction. The Repository implements data synchronization and conflict handling of the two `DataSource`s. I haven't studied the `RxDB` you mentioned in detail, but I believe that even if you change to any other local database, you can provide a `LocalDataSource` implementation, thereby decoupling the implementation of the local database from other components in the App. Even if `RxDB` or any local database implementation is unavailable again for some reason in the future, all we need to do is replace a `LocalDataSource` implementation.

The main purpose of `RemoteDataSource` is data synchronization. We previously synchronized data directly through Realm and MongoDB Atlas App Service. My suggestion is that we implement a data backend web application based on RESTful API and use production-level database programs to store data in this web application. We can also synchronize attachments, including Paper's PDFs and other attachments, uniformly through this backend.

I am happy to contribute an independent Web API application for synchronization backend. This web service is designed to run on the cloud, interact with the application through the HTTP-based Restful API, and provide a relatively simple web page to view the content in the database online (of course, the plug-in system cannot be implemented in the Web, but the client can store the data processed by the plug-in system in this API backend). We can provide a version that users can manage by themselves. Users with certain operation and maintenance experience can host their own storage backend. I can also develop an extended version to provide official synchronization services from Paperlib. This extended version will expand the account system belonging to Paperlib and lay the foundation for future team sharing and other functions.

The API backend idea is still in the prototype stage, but if it works (in other words, Paperlib has ambitions to offer official services), I'd be happy to discuss the details. Let me know what others think of this part.

**Reply by GeoffreyChen777** [reply]
I investigate the RxDB, and I think their implementation is very similar to your design. The biggest problem of RxDB is that without the paid plan, we have very few options for the backend database. Even the sqlite database backend is provided by the premium plan.

Their design is:
- we write the data to a local database first, and we pull the newest version of the server database to the local. If we there is no conflict, merge the local with the server version and push the merged version to the server.
- if there is any conflict, we handle the conflict on the local client and then push the merged version to the server.
- there is a server event to tell each client that there are new changes. Once the client receives a server event, it pull the new data from the server.
- there are many options for the local/server database backend, they implement a layer between the real database backend and the application to handle data query, and sync.
- they also have an HTTP RestfulAPI server to communicate between the client app and server database. In addition, they also have other options such as WebSocket etc.

For your design, it sounds good. We can release the online database sync service as a package or a docker image. For professional users, they can deploy their own backend on their private cloud, NAS etc. We can also deploy it on our cloud to provide an official server in the future.

**Reply by GeoffreyChen777** [reply]
I think the best solution is we implement the database infrastructure by ourselves. We can build a layer based on popular production-level databases such as sqlite, MongoDB, Mysql etc. as your design.

In this way, we will never be controlled by others again. I fear that databases like RxDB might one day turn some important features into paid ones.

**Reply by charlieJ107** [reply]
It sounds like we don't have a lightweight database to process data locally in Paperlib except Sqlite. In fact, although Paperlib uses data structures such as Json objects to process data internally, Json objects have almost unchanged schemas, which means that document-based databases are not necessary. Through some simple serialization, we can still save data in lightweight local databases such as Sqlite. If Sqlite can be used, at least we don't need to run a bloated database management program like Mysql on each user's device.

If SQL database is not a good idea, I have a more violent way to solve the problem of local databases. All local data is directly stored as Json files, that is, no database is used, only JSON files are used as local data storage. When syncing, the content of the JSON file can be sent to the backend according to the design of the API.

My current plan is that we still use a mature SQL database such as Postgresql for the backend. The consideration for this is that it is easy to deploy on different platforms. As for the local, in the case of synchronization (I believe most users want it), it is just a local cache of data. If we design the official synchronization to be simple and easy to use, more users will be more inclined to use the synchronization function.

I think the local-first strategy is more for privacy and portability for ordinary users, so we can integrate the online synchronization service into Paperlib as an important function.

By the way, I mentioned earlier that I want to refactor the entire application. One of the very important reasons is that I want to take this opportunity to split Paperlib into three more independent parts (the "independent" here means that each of them can be distributed, deployed and run as an independent program or package). In my design, these three parts are:
-  The UI is a completely independent Vue single-page application (SPA), which can be hosted on the cloud as a web page, run in the user's browser, or used in a browser-based desktop app such as Electron.
- The data backend, including the Scrapper and data storage backend, interacts with the UI through APIs based on Http or other protocols. It can run in the service process of Electron or in a cloud environment as a Web API service.
- The plugin system is provided in desktop applications based on Electron. It interacts with the UI through an interface similar to PLAPI used in previous designs. It is equivalent to a special UI and is also connected to the data backend through an API.

The idea of ​​this design is still very rough, especially the interaction between the plugin system and the UI. My original idea was to completely reconstruct the underlying layer of the existing plugin system so that the plugin system runs as a rendering process like the UI instead of a VM running in a service process. There are two considerations involved here:

- In the previous communication, you mentioned that the main reason for not choosing to design the plugin to run in the rendering process was performance (mainly memory usage) concern. I have done some simple tests. When the plugin system does not need to stay in the background, the performance occupation is actually limited. Considering that the existing plugin needs to register hooks, we can design the plugin to start a rendering process only when needed and close the rendering process after the operation ends, which can greatly reduce the long-term memory occupation.
- Another consideration is actually for maintainability and security considerations. Currently, plugins are hosted and distributed on the npm platform in the form of npm packages. On the one hand, this will occupy npm's public resources, and on the other hand, it will make it difficult to maintain our plugin market. Imagine that an attacker maliciously submits a large number of plugins that conform to the paperlib plugin naming format to npm, and we cannot effectively filter and control the plugin list. The plugin market will be a disaster for users. If the plugin is allowed to run in the rendering process, the UI required by the plugin can be provided by the plugin developer, and Paperlib's UI does not need to set slots for a variety of plugins. At the same time, we can set up a centralized plugin market to facilitate our management of plugins. Most of the existing plugins do not involve the differences between the Node environment and the browser environment. For most existing plugins, they can be moved from the current VM running environment to the browser environment of the rendering process without modification. Before we need to face the technical debt brought by a large number of plugins, it may be a better choice to make modifications early.

**Reply by GeoffreyChen777** [reply]
1. **JSON**: we cannot use JSON as the local database because we need a good query performance. For the Realm, we have the realm query language and the `filter()` API. For SQLite, we have SQL. We cannot use JSON files to store the data and implement our own query engine.
2. **For the server database**. it's ok to choose any mature database. We can choose one, such as Postgresql as a start, and abstract an adapter layer. Later we can introduce more options with different adapter layers.
3. **For the local database**. I have two options:
  - Use SQLite. I'm open to it. I googled the lightweight NOSQL dataset. The answer is that SQLite is the best. If we choose SQLite as the local database, we need to:
    - implement a function to transform the realm query language into SQL. as there are many users who have smart filters.
    - reimplement the CRUD.
  - Still use Realm as the local database. They mentioned that Realm.js will still be available as an opensource database. So we can use it without reimplementing the query transform and the CRUD. We only need to implement a sync server.
4. **About local first policy**. It's not only about privacy. In your design, it seems that the server database is the master. The local database is just a cache of the server database. In my opinion:
  - the local database should be the master, which means we have several masters.
  - the server database serves as a backup, and a communication centre for masters to sync data.
  - one user will only use one device at the same time. and many users including me require the offline mode. Because sometimes I need to modify the local database without the internet.
  - we should consider that many users only want to use the local database. They may only have one device. They don't want to give their data to our server, and they don't know how to set up a cloud server.
  - we should do the query in the local database. It's faster than querying a server database.
5. **About splitting Paperlib into three more independent parts, deployment, and distribution**: your design is cool, I agree part of them. My main concern is that we should consider the balance between the workload and values:
  - You and I are both PhD students, I believe we don't have too much time to work on Paperlib.
  - We should consider the application scenarios. I think we only have three use cases:
    - The desktop Electron version, which is the most important one.
    - The web page version in future. If a user wants to do some lightweight tasks on a new device. They can use the webpage version. As I have already split the front- and back- end in the new branch, implementing a webpage version should be easy. We can let all services run in the node, run an HTTP server for the UI, and implement a new RPC protocol to use WebSocket to communicate between them. We can distribute the webpage version as a package or a docker image. Users can deploy it on their own server.
    - The iPad version in future. As we don't have time to implement the iPad version. Maybe users can use the webpage version as a workaround.

    As you can see, we don't have to refactor the whole structure. Your fully separated design is cool, but may introduce limited benefits to users. And it's too risky.
6. **Extension system** I agree the current extension system is suboptimal. But I don't get it why do you want all extensions run in individual renderer processes? Currently, we have an extension type called `new window extension`, it can create a new window in an individual renderer process, show their own UI etc. For example, the `AIChat` extension, it create a chat window. For simple extensions such as the `citation-count`, it's just simple code snippets running in the VM. Simple extensions don't need the renderer process at all. The creation of a renderer process is slow. we cannot open it, close it, and open it again. Imagine, for the `citation-conut` extension, every time we choose a different paper, we get the citation count from the internet and show it in the details panel. We cannot put it in a renderer process. It's too slow. In the current design. Simple extensions run in the VM, Complex extensions can create their own renderer processes.
7. **Extension Market**. yes I agree that there is a risk that bad guys may pollute our market. If we want to avoid this, we need to create our own extension repository and introduce the review process. I think it's hard for our small team. Maybe in the future, we can do it in the future. But now I think we should focus on the database issue.

**Reply by charlieJ107** [reply]
I completely agree with your point about architecture refactoring. I just take this opportunity to throw out my previous very rough ideas for discussion, and I have no intention of promoting the refactoring of the entire architecture. I believe that many issues about workload (such as extension market) will be gradually resolved in the future. We can of course focus on data storage issues now.

Regarding local first policy, I agree that the pure local mode should be the main data storage mode, and the online service should be used as the role of data backup and synchronization. What I mean by local cache is that after the user edits the local data, once the network is connected, the data can be synchronized to the backend so it can be synchronized to other devices. Users can of course choose to always run locally as long as the synchronization function is not enabled. This is actually a UI/UX issue, that is, whether the software encourages users to use network services in UI/UX design, or only provides corresponding functions when users actively seek synchronization services.

Regarding the adaptation of Smart filter and SQLite, we can solve it with the help of mature ORM. In fact, in the previous implementation based on Realm, I noticed that our smart filter query is actually spliced ​​with the corresponding Realm query statement. In SQL database, we can of course also implement it by splicing SQL statements. A more professional approach is to use a database operation library, such as ORM, to map the Smart filter configuration to a piece of ORM query code, thereby realizing the function of the smart filter. Although this requires us to re-implement the query part of the service, it should not be too difficult. Some of the famous NodeJS ORMs I know are Sequelize (https://sequelize.org) and Prisma (https://www.prisma.io), of course there are many other options. Perhaps we can consider implementing the corresponding functions based on a certain ORM.

**Reply by charlieJ107** [reply]
I would like to add one thing about the Local first policy. We mentioned earlier that we hope that changes on different devices can be synchronized to other devices immediately. Generally speaking, no matter what SDK is used, there are essentially only two ideas:

  1. Use an online event publishing/subscribing service. When device A changes, device A sends an event to the online service, and the online service pushes an event notification to all clients that have subscribed to the change event. For other clients, it is necessary to receive events from the online service in a long connection or polling manner. The advantage of this is that changes are timelier and can provide a seamless synchronization experience. But for developers, it would be a significant cost to either write or maintain such a synchronization infrastructure, including client and server components. The performance consumption of this part is also very high. If the design is relatively rough, it may affect the software experience. Therefore, many similar services use native libraries like those developed in C/C++ to operate on the synchronization components. I think this is also why MongoDB Atlas deprecated their Device Sync, service and RxDB puts many important functions in paid projects.

  2. Use a unique authoritative online data source. All changes made by the client while offline will be cached locally, and these changes will be sent to the online service when connected to the online service, and the local data list will be updated based on the (changed) data of the online service. The advantage of this is that there is no need to consider maintaining the event push service. The design of the online synchronization data source is relatively simple. You only need to implement a full-featured API. The client can use protocols such as HTTP to implement functions with the online service. The disadvantage is that this means that once the user turns on online synchronization, the online data end is essentially used as the authoritative data source, and the local client will follow the online data as much as possible. In fact, in the era of Realm, we also pushed data to the database on the Atlas server through Realm, and then used the event push function provided by Device Sync to require all connected clients to pull data from the Atlas server. Our approach is simply to turn the passive operation of event push into an active request every time the client loads the database. Assuming we add some mechanisms, such as triggering a request when a component is rendered (this request can be made asynchronously in a separate process without blocking the rendering of the UI), we can also achieve (although not truly "real-time") fairly timely synchronization.

The considerations in the local first policy you mentioned earlier are not a problem when running purely locally, because the client (i.e. Paperlib APP) of course stores and operates data in the local SQLite (or other database). If users need to synchronize data but do not trust the official synchronization service of Paperlib, users can of course choose to achieve it through self-hosted independent data synchronization backend. We can provide some quick operation options, allowing users to deploy to platforms such as Vercel, Cloudflare, DigitalOcean with one click. These operations will not be more complicated for users who have the ability to purchase MongoDB Atlas services in previous operations. As for users with more stringent data security requirements, I think manually deploying our synchronization backend to their own servers will not be a big problem for them. In short, users always have a choice. If you want to save trouble and zero learning cost, just use the official service of Paperlib; if you have higher requirements, choose the corresponding solution according to your ability.

**Reply by GeoffreyChen777** [reply]
Hi, thanks for your introduction.

I think, for paperlib, we don't need a real realtime sync database. We can tolerate some latency.

If I understand correctly, RxDB they use the `server sent event` to send signal to client APPs.
https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events

I guess it's a lightweight solution?

Your design sounds ok. We can start from it.

And about the deployment of the server database. Yes that's what I want. Professional users can deploy their own server. Other users can use the official server, maybe we can also provide free and paid plan in the future.

**Reply by charlieJ107** [reply]
Our current Sync API uses a RESTful API to implement data interaction, and currently considers polling to obtain database updates. Regarding the specific update method, I just thought of a way for the client to know whether it needs to obtain updated data and refresh the rendering. I will provide a Diff API in the API, which the client can use to call Get (perhaps with a local Diff ID), and the server will return the latest Diff ID. When the Diff ID carried in the request does not match the latest ID on the server (the server data has been updated), the Diff API will also return all the data that has changed since the requested ID and the server's latest ID (including data types, such as updated Paper or updated Tag) and the ID of the updated data entry. This way, the client can actively send a Get request again to get the corresponding updated data according to the ID in turn. I'm still wondering whether to set up a separate field for Diff to indicate whether new data has been created, so that the client can actively refresh the entire table; the alternative is to let the client determine whether to refresh the entire table based on the Diff obtained and by checking its own database records. But I think this Diff API can be used to obtain data updates in a lightweight manner, so that the client can update the data on demand without having to refresh the database each time it polls to trigger UI rendering.

**Reply by GeoffreyChen777** [reply]
> the Diff API will also return all the data that has changed since the requested ID and the server's latest ID (including data types, such as updated Paper or updated Tag) and the ID of the updated data entry.

seems that you need to maintain a log to record all changes to the server db right?

**Reply by charlieJ107** [reply]
Yeah, some sort of that. It can also be implement in other backend tech stack (say some cloud native log service) but add a table in DB will be more easier to implement so we can make our sync service available asap. Since the diff API is unlikely changing if we changed the backend implement we can transfer from DB table to a more demure way later. Currently I'm planning to add a middleware that create a new record to the diff table whenever a write operation is made to the DB.

**Reply by GeoffreyChen777** [reply]
Cool, I think with this log DB, we can even implement more detailed conflict handling in the future.

**Comment by GeoffreyChen777** [comment]
I just checked the atlas, currently we cannot even create a new sync service. It means that for all new users, Paperlib cannot do the cross device sync.

**Comment by ZanderZhao** [comment]
Can I deploy MongoDB Atlas Sync in the cloud by myself, what should I fill in that address, I use docker `mongodb/mongodb-atlas-local`, and I found that filling in the ip+ port is not working

**Reply by GeoffreyChen777** [reply]
It’s not the sync server. Currently there is no way to self host a MongoDB sync server. We are developing our own sync server. Just wait for it.

**Comment by charlieJ107** [comment]
## 🚀 Paperlib Official Sync Service Update & Call for Contributors!  

Hello everyone! 👋  

We’re excited to share the latest progress on **Paperlib Official Sync Service**! After extensive development (includes rebuilding from scratch 2 times), we have built a simple and rough synchronization architecture, including a **remote sync API** and a **local sync service** integrated into the client application. We have also deployed a **sync service dev server** for development and testing (many thanks for our honorable sponsor [DigitalOcean](https://www.digitalocean.com)).  

### 📌 Current Progress:
✅ Initial implementation of the remote sync API  
✅ Local sync service integrated into the client  
✅ Basic data synchronization workflow completed  
✅ Authentication and authorization handled via Auth0  
✅ Internal test server deployed  

### 🧪 Looking for Developers to Test & Contribute:  
The sync functionality is still **under development and testing**, and we are actively improving it in the **chore/official-sync** branch. We encourage developers to **checkout this branch** to test and experiment with the new sync feature. **Please note: This feature is still in the developing phase and insanely unstable—do not rely on it for important data!**  

We’re especially looking for contributors with **Electron / Node.js** experience to help us:  
🔹 Test and debug multi-device synchronization scenarios  
🔹 Validate sync robustness in complex network environments  
🔹 Provide bug reports & performance improvement suggestions  
🔹 Contribute code to enhance the synchronization mechanism  

### 📢 How to Get Involved:  
If you're interested in testing or contributing, feel free to:  
🔹 Share your feedback and suggestions in our **GitHub Discussions**  
🔹 Join our **Discord community** to discuss with the team  
🔹 Submit PRs to help improve the sync service 🚀  

Your contributions will help make **Paperlib Sync Service** more stable and reliable, allowing users to enjoy seamless synchronization across devices! 🎉  

Thanks for your support, and we look forward to working together to make Paperlib even better! 💙
